spec/android_unit_test_pr_checker_spec.rb

```diff
@@ -0,0 +1,222 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+
+module Danger
+  describe Danger::AndroidUnitTestPRChecker do
+    it 'should be a plugin' do
+      expect(described_class.new(nil)).to be_a Danger::Plugin
+    end
+
+    describe 'with Dangerfile' do
+      before do
+        @dangerfile = testing_dangerfile
+        @plugin = @dangerfile.android_unit_test_pr_checker
+
+        allow(@plugin.github).to receive(:pr_labels).and_return(['my_label'])
+      end
+
+      it 'shows the right number of errors when a PR adding new classes that do not have corresponding tests' do
+        changes_dict = {
+          'Abc.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { println(""); } }',
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { fun testMethod() {} }',
+          'Polygon.kt' => 'abstract class Polygon { abstract fun draw() }',
+          'Abcdef.kt' => 'class Abcdef(name: String) { fun testMe() { println("") } }',
+          'TestsINeedThem.java' => 'public final class TestsINeedThem { public void testMe() { System.out.println(""); } }',
+          'TestsINeedThem2.kt' => 'public open class TestsINeedThem2 { fun testMe2() { } }'
+        }
+
+        diff = generate_add_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests
+
+        expect(@dangerfile.status_report[:errors].count).to eq 5
+      end
+
+      it 'does not show errors when new classes have corresponding tests' do
+        changes_dict = {
+          'Abc.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { System.out.println(""); } }',
+          'project/src/androidTest/java/org/test/AbcTests.java' => 'class AbcTests { public void abcTest() { sut.main([]) } }',
```

Comment:
I don't think this should be considered a test for the class `Abc`?
While the test class is called `class AbcTests`, none of the code in that test file references the `Abc` class (like `Abc.main([])`.

That is, except I'm missing some Java magic that I'm not aware of (which is possible, as I never wrote any serious Java personally…) which would make the `sut` variable magically be defined by the Java compiler on test files as being the class with the same name as the test class `AbcTests` but without the `Tests` suffix…? Which is what would make that Java code compile and be considered a test on the `Abc` class itself?

Reply:
That comes from the misunderstanding you pointed out in [this suggestion](https://github.com/Automattic/dangermattic/pull/1#discussion_r1260934060).

I'll change these tests that have no usage at all (no magic in Java to auto-create a `sut`, unfortunately 🪄😅) to reflect that.